### PR TITLE
Jg/transaction state

### LINKED ISSUE #2551: Add State property to Transaction
At the moment it's not possible to know what is the state of a certain transaction, and the developer needs to be careful not to commit or rollback the transaction twice in order to avoid exceptions

### PR DESCRIPTION
**Pull request by JacobOscarGunnarsson**
## Description
Allows a user to check the state of a transaction so it doesn't mistakenly commit/rollback twice

Fixes #2551

### CODE REVIEW

File: Tests/Realm.Tests/Database/InstanceTests.cs [hunk 1]
```diff
@@ -199,6 +199,22 @@ public void InstancesHaveDifferentHashes()
             Assert.That(realm1.GetHashCode(), Is.Not.EqualTo(realm2.GetHashCode()));
         }
```
Comment (JacobOscarGunnarsson) [inline Tests/Realm.Tests/Database/InstanceTests.cs:201]: Not entirely sure if this is the correct place to put the test

**Review by papafe — CHANGES_REQUESTED** [review]

**Comment by papafe** [comment]
You should also check the CI ;) 
If something is not clear we can talk about it! The UI is a little messy

File: Realm/Realm/Transaction.cs [hunk 2]
```diff
@@ -23,6 +23,28 @@
 
 namespace Realms
 {
+        /// <summary>
+        /// Represents the state of a <see cref="Transaction"/>.
+        /// </summary>
+    [Flags]
```
Comment (nirinchev) [inline Realm/Realm/Transaction.cs:29]: This should probably not be flags - MS has pretty good docs about what `[Flags]` designates and when to use it [here](https://learn.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=net-7.0).

File: Realm/Realm/Transaction.cs [hunk 3]
```diff
@@ -59,6 +88,7 @@ public void Rollback()
         {
             EnsureActionFeasibility("roll back");
             _realm.SharedRealmHandle.CancelTransaction();
+            State = TransactionState.RolledBack;
             FinishTransaction();
```
Comment (nirinchev) [inline Realm/Realm/Transaction.cs:92]: We can probably move setting the state to `FinishTransaction` and then pass the value as an argument. Not saving a whole lot of code, but it's a bit more expressive as "this is the final thing we need to do as we close out the transaction".

File: Realm/Realm/Transaction.cs [hunk 4]
```diff
@@ -23,6 +23,27 @@
 
 namespace Realms
 {
+        /// <summary>
```
Comment (nirinchev) [inline Realm/Realm/Transaction.cs:26]: Spacing seems off here.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,7 @@
 ## vNext (TBD)
 
 ### Enhancements
+* Added a field `Transaction.State` which describes the current state of the transaction.
```
Comment (LaPeste) [inline CHANGELOG.md:4]:
```suggestion
* Added a field `Transaction.State` which describes the current state of the transaction. (Issue [#2551](https://github.com/realm/realm-dotnet/issues/2551)
```

**Review by LaPeste — CHANGES_REQUESTED** [review]
Just some very minor details. Otherwise, the work loos fine.

**Comment by papafe** [comment]
Looks good! I would just add one test, to verify that it works also for `CommitAsync` and then if the CI agrees, it's good to go ;)

**Pull request merged**